tests/virt_autotest/sriov_network_card_pci_passthrough.pm

```diff
@@ -0,0 +1,346 @@
+
+        #check host and guest to make sure they work well
+        my $check_host_result = check_host();
+        my $check_guest_result = check_guest($guest);
+        $self->{test_results}->{$guest}->{"Error: host '$check_host_result' after the first vf hot unpluged from guest!"}->{status} = 'FAILED' if $check_host_result ne "Fine";
+        $self->{test_results}->{$guest}->{"Error: guest '$check_guest_result' after the first vf hot unpluged from guest!"}->{status} = 'FAILED' if $check_guest_result ne "Fine";
+
+        #check again the remaining vf(s) inside vm
+        for (my $i = 1; $i < $passthru_vf_count; $i++) {
+            test_network_interface($guest, gate => $gateway, mac => $vfs[$i]->{vm_mac}, net => 'br123');
+        }
+        set_var("SRIOV_NETWORK_CARD_PCI_PASSSHTROUGH", 0);    #turn off the flag in case of affecting other tests
+
+        #unplug the remaining vf(s) from vm
+        for (my $i = 1; $i < $passthru_vf_count; $i++) {
+            unplug_vf_from_vm($guest, $vfs[$i]);
+            record_info("VF unplug from vm", "$vfs[$i]->{host_id} \nGuest: $guest \nbdf='$vfs[$i]->{vm_bdf}'   mac='$vfs[$i]->{vm_mac}'   nic='$vfs[$i]->{vm_nic}'");
+            assert_script_run("virsh nodedev-reattach $vfs[$i]->{host_id}", 60);
+            record_info("Reattach VF to host", "vm=$guest \nvf=$vfs[$i]->{host_id}");
+        }
+        script_run "lspci | grep Ethernet";
+        save_screenshot;
+  
+        #check host and guest to make sure they work well
+        $check_host_result = check_host();
+        $check_guest_result = check_guest($guest);
+        $self->{test_results}->{$guest}->{"Error: host '$check_host_result' after vfs hot unpluged from guest!"}->{status} = 'FAILED' if $check_host_result ne "Fine";
+        $self->{test_results}->{$guest}->{"Error: guest '$check_guest_result' after vfs hot unpluged from guest!"}->{status} = 'FAILED' if $check_guest_result ne "Fine";
+
+    }
+
+}
+
+
+#set up ssh, packages and iommu on host
+sub prepare_host {
+
+    #install required packages on host
+    zypper_call '-t in pciutils nmap';    #to run 'lspci' and 'nmap' command
+
+    #check IOMMU on XEN is enabled
+    if (get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen")) {
+        assert_script_run "xl dmesg | grep IOMMU | grep -i enabled";
+    }
+}
+
+
+#get the BDF the PF device on host
+sub get_sriov_ethernet_devices {
+
+    #get the BDF of the ethernet devices with SR-IOV
+    my $nic_devices = script_output "lspci | grep Ethernet | grep -v 'Virtual Function' | cut -d ' ' -f1";
+    my @nic_devices = split("\n", $nic_devices);
+    my @sriov_devices;
+    foreach (@nic_devices) {
+        if ((script_run "lspci -v -s $_ | grep -q 'SR-IOV'") == 0) {
+            push @sriov_devices, $_;
+        }
+    }
+    record_soft_failure "Error: there is no ethernet devices with SR-IOV in the host and test aborts!" if !$#sriov_devices;
```

Comment:
Please record soft failure in main body, not inside individual functions. And, at the same time, add corresponding junit log soft failure item.

Reply:
changed along with another fix.